sdk/iot/azure-iot-modelsrepository/azure/iot/modelsrepository/client.py

```diff
@@ -0,0 +1,173 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+import six.moves.urllib as urllib
+import re
+from azure.core import PipelineClient
+from azure.core.pipeline.transport import RequestsTransport
+from azure.core.configuration import Configuration
+from azure.core.pipeline.policies import (
+    UserAgentPolicy,
+    HeadersPolicy,
+    RetryPolicy,
+    RedirectPolicy,
+    BearerTokenCredentialPolicy,
+    ContentDecodePolicy,
+    NetworkTraceLoggingPolicy,
+    ProxyPolicy,
+)
+from . import resolver as resolver
+from . import pseudo_parser
+from .chainable_exception import ChainableException
+
+
+# Public constants exposed to consumers
+DEPENDENCY_MODE_TRY_FROM_EXPANDED = "tryFromExpanded"
+DEPENDENCY_MODE_DISABLED = "disabled"
+DEPENDENCY_MODE_ENABLED = "enabled"
+
+
+# Convention-private constants
+_DEFAULT_LOCATION = "https://devicemodels.azure.com"
+_DEFAULT_API_VERSION = "2021-02-11"
+_REMOTE_PROTOCOLS = ["http", "https"]
+
+
+class ModelsRepositoryClient(object):
+    """Client providing APIs for Models Repository operations"""
+
+    def __init__(self, repository_location=None, api_version=None, **kwargs):
```

Comment:
Something to consider when no `repository_location` is given.  In this case we default to the global IoT models repository endpoint and here we can guarantee an expanded model file for any given model is available.

The C# implementation supports an optional model dependency resolution option through its initialization/ctor's client options, but if you don't provide one and are pointing to the global endpoint the default model dependency resolution becomes the try from expanded mode.

Reply:
To be clear, the desired design is to have a dependency mode tied to the instance of the client, that can be overridden when using the `.get_models()` API by passing in an optional argument?

I do think this relationship (especially when factoring in the elsewhere-mentioned dynamic default dependency mode based upon target URI) could be somewhat confusing and difficult to capture in documentation, but I can implement that if it's the requirement.

Reply:
I was addressing the other https://github.com/Azure/azure-sdk-for-python/pull/17180#discussion_r591987457 around this topic but will continue here and what you said is the gist of it.

1. Have an optional dependency resolution mode in client initialization
2. If no repository location is chosen, we default to the global endpoint. If we are pointing to the global endpoint and no dependency resolution is given on init, we default to try from expanded mode because the service will have these types of files.
3. If any other repository location is given the default dependency resolution mode is enabled.
4. Like you said `.get_models()` has an optional parameter which will override the client instance dep mode.

The samples here tries to explain the dependency resolution when pointing to the global endpoint https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/modelsrepository/Azure.Iot.ModelsRepository/samples#initializing-the-models-repository-client